### ISSUE TITLE
A11y keyboard focus control in the Slide Toggle Component

### ISSUE THREAD

**Issue by DaveBest99** [angular/material2#3591]
#### Bug, feature request, or proposal:
Bug:
The Tab key and Shift+Tab key should move focus to and and away from the Radio button component, and the UP/Down arrow keys should rotate focus between each Radio button in the group. Only the selected Radio button in the group gets focus.

#### What is the expected behavior?
Arrow keys are expected to rotate between Radio buttons, and Tab key moves focus away from the group.

#### What is the current behavior?
Currently the Tab key and Arrow key focus order is unexpected.

#### What are the steps to reproduce?
See below.

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Plunker template: https://goo.gl/DlHd6U

#### What is the use-case or motivation for changing an existing behavior?
To create a consistent and accessible experience for all users.

#### Which versions of Angular, Material, OS, browsers are affected?
Using JAWS17 screen reader with browsers Firefox, Chrome, and IE11.

#### Is there anything else we should know?
The screen reader user experience follows:

1. Press Tab key to the Radio button component.
JAWS says "Accent radio button checked 2 of 3", and the braille display shows "< > Accent rbtn 2 of 3".
Note, the Up and Down arrow key moves freely by the Radio buttons and page content, and the user can press Spacebar to check a specific Radio button, but the arrow keys do not rotate between just the buttons as expected.
Also, the text "Color" is not associated with the Radio buttons, and thus not spoken or shown on the braille display.

2. Press Tab key again.
JAWS says "Checked check box not checked", and the braille display shows "< > Checked chk".
Pressing Spacebar will Check and Uncheck this element with the appropriate response spoken by JAWS, but there is no apparent effect as the result does not change and there is no aria-alert.

3. Press Tab key again.
JAWS says "Disabled check box not checked", and the braille display shows "< > Disabled chk".
Pressing Spacebar will Check and Uncheck this element with the appropriate response spoken by JAWS. When not Checked the Tab key moves focus to the "Slide me!" element, and when Checked the "Slide me!" element is skipped over. This behaviour is inconsistent with other Angular components.

4. Press Tab key to move focus to the "Slide me!" element.
JAWS says "Slide me! check box not checked", and the braille display shows "< > Slide me! chk".
Pressing Spacebar will Check and Uncheck this element with the appropriate response spoken by JAWS. There is no apparent change to the screen reader user when Checked or Unchecked (no aria-alert).
The Input element has no user instructions, or aria label, to describe the purpose as a Slide Toggle.

5. When using IE11 each element appears in the focus order twice and the text labels are improperly aligned.
Example, pressing Tab key to move focus to the Result "Slide me!" element JAWS says "Angular check box not checked", and shows "< > Angular chk" on the braille display.

**Comment by DevVersion** [comment]
Hi @DaveBest99. I need to ask, why did you mention `radio buttons` for the `slide-toggle`?

The `slide-toggle` should work like a native checkbox (we are extending the "native" accessibility).

**Note**: The slide-toggle does not support any group of elements.

**Comment by DaveBest99** [comment]
Good question!

I am testing the Angular Material demo Slide Toggle Component

https://material.angular.io/components/component/slide-toggle

 

from a screen reader user perspective, and have not looked at the code. However, in looking at the code, the demo is confusing, as I do not see Radio buttons in the md-slide-toggle module.

 

At the top of the Tab Panel example is the MD-CARD-CONTENT, which encloses:

- Heading "Slider configuration" (h2 class=example-h2)

- MD-RADIO-GROUP with role=radiogroup, which contains 3 Radio Buttons (MD-RADIO-BUTTON).

- MD-CHECKBOX with 2 Input elements with type=checkbox (labelled Checked and Disabled).

 

At the bottom of the Tab Panel example is the SLIDE-TOGGLE-CONFIGURABLE-EXAMPLE, which encloses:

- Heading "Result" (h2 class=example-h2).

- MD-SLIDE-TOGGLE with 1 Input element type=checkbox.

- MD-CARD-CONTENT and MD-CARD.

 

Question:

1. Is the top part of this Tab Panel part of the md-slide-toggle module testing?

2. The md-slide-toggle is an on/off control that can be toggled via clicking or dragging (wich can be toggled with the Spacebar, but other than the Disabled checkbox what effect should the Radio Buttons have on this "Result" element)?

 

Hope this helps.

David

**Comment by DevVersion** [comment]
@DaveBest99 Exactly, the only thing that is part of the `md-slide-toggle` is the output from the `slide-toggle-configurable-example`.

Can you please re-test with the following page: [here](http://plnkr.co/edit/I8NPCFpn1btNkL4R6osN?p=preview)

**Comment by DaveBest99** [comment]
Paul, the Plunker website is very screen reader unfriendly, and I was unable to assess the demo. However, I will work on it and get back to you.

David

**Comment by DaveBest99** [comment]
Paul, I have consulted with colleagues about the Angular Slider Toggle demo. The consensus is that the purpose of the slider is unclear in the demo. If it is simply to Check and Uncheck a Checkbox, then there is no accessibility issue, as the Checkbox is keyboard operable. However, the text label “Slide me” is misleading and confusing to not only blind screen reader users.

 

If the visual slide action of the Checkbox is to portray important information, then

A)     The slide action should include an aria-alert to indicate screen change of Slide Up Checked or Slide Down Unchecked,

Or

B)      It may be more appropriate to implement a musical ascending tone when Checked and a Descending musical tone when Unchecked. Tones can be used in place of text labels and speech output when appropriate.

 

Hope this is helpful.

David

 

From: David Best [mailto:davebest@cogeco.ca] 
Sent: March 27, 2017 03:40 PM
To: 'angular/material2'; 'angular/material2'
Cc: 'Mention'
Subject: RE: [angular/material2] A11y keyboard focus control in the Slide Toggle Component (#3591)

 

Paul, the Plunker website is very screen reader unfriendly, and I was unable to assess the demo. However, I will work on it and get back to you.

David

**Comment by DevVersion** [comment]
Thanks @DaveBest99. Yes it's more likely a `checkbox` just with an extra option that allows people to change the `checked` state using gestures.

I understand that the label of the checkbox is kind of confusing but I think it is just to show that you can also slide it.

I'm closing this issue as I think the a11y works fine on the slide-toggle (it's actually a native checkbox).

Thanks for the issue and details!
